deckhouse-controller/pkg/controller/moduleloader/loader.go

```diff
@@ -0,0 +1,486 @@
+// Copyright 2024 Flant JSC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package moduleloader
+
+import (
+	"context"
+	"fmt"
+	"io/fs"
+	"maps"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/flant/addon-operator/pkg/module_manager/loader"
+	"github.com/flant/addon-operator/pkg/module_manager/models/modules"
+	"github.com/flant/addon-operator/pkg/utils"
+
+	"gopkg.in/yaml.v3"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/deckhouse/deckhouse/deckhouse-controller/pkg/apis/deckhouse.io/v1alpha1"
+	"github.com/deckhouse/deckhouse/deckhouse-controller/pkg/helpers"
+	d8config "github.com/deckhouse/deckhouse/go_lib/deckhouse-config"
+	"github.com/deckhouse/deckhouse/go_lib/deckhouse-config/conversion"
+	d8env "github.com/deckhouse/deckhouse/go_lib/deckhouse-config/env"
+	"github.com/deckhouse/deckhouse/go_lib/dependency/extenders"
+	"github.com/deckhouse/deckhouse/pkg/log"
+)
+
+const (
+	moduleOrderIdx = 2
+	moduleNameIdx  = 3
+)
+
+var (
+	// some ephemeral modules, which we even don't want to load
+	excludeModules = map[string]struct{}{
+		"000-common":           {},
+		"007-registrypackages": {},
+	}
+
+	// validModuleNameRe defines a valid module name. It may have a number prefix: it is an order of the module.
+	validModuleNameRe = regexp.MustCompile(`^(([0-9]+)-)?(.+)$`)
+)
+
+var _ loader.ModuleLoader = &Loader{}
+
+type Loader struct {
+	client         client.Client
+	log            *log.Logger
+	embeddedPolicy *helpers.ModuleUpdatePolicySpecContainer
+	version        string
+	modulesDirs    []string
+	modules        map[string]*Module
+}
+
+func New(client client.Client, version, modulesDir string, embeddedPolicy *helpers.ModuleUpdatePolicySpecContainer, logger *log.Logger) *Loader {
+	return &Loader{
+		client:         client,
+		log:            logger,
+		modulesDirs:    utils.SplitToPaths(modulesDir),
+		modules:        make(map[string]*Module),
+		embeddedPolicy: embeddedPolicy,
+		version:        version,
+	}
+}
+
+// LoadModules implements the module loader interface from addon-operator, used for registering modules in addon-operator
+func (l *Loader) LoadModules() ([]*modules.BasicModule, error) {
+	result := make([]*modules.BasicModule, 0, len(l.modules))
+
+	for _, module := range l.modules {
+		result = append(result, module.GetBasicModule())
```

Comment:
what if basic module is nil?

Reply:
basic module is set when module is created